contributors/design-proposals/storage/postpone-pvc-deletion-if-used-in-a-pod.md

```diff
@@ -33,37 +33,33 @@ Depending on the storage type the data loss occurs in one of the below scenarios
-
-Note: as soon as all the PVC(s) with `deletionTimestamp` set are deleted (in case none of the PVC(s) don't have finalizer information set) the pod that now uses non-existent PVC(s) will be started and will be in pending state (waiting for the non-existent PVC(s) to be created). Otherwise, in case at least one PVC has finalizer information set, i.e. PVC deletion controller is waiting for the scheduled pod to be terminated, there will be a deadlock: kubelet waiting for PVC deletion while PVC deletion controller waiting for pod termination. The deadlock can be resolved only by deleting the scheduled pod.
+In case any of the PVC(s) used by the pod has the `deletionTimestamp` set kubelet won't start the pod but will report and error: "can't start pod (%pod) because it's using PVC (%pvcName) that is being deleted". Kubelet will follow the same code path as if PVC(s) do not exist.
 
 ### PVC Deletion Controller
 PVC deletion controller is a new internal controller.
 
-PVC deletion controller can list pods and watch for pods. It caches status of pods that use a PVC. It also watches for PVC deletes.
-
-PVC deletion controller reacts to PVC delete by doing live update of pods in the PVC's namespace and afterwards does one of the below:
-- In case the PVC is not used by a pod: removes the finalizer information from the PVC metadata.
-- In case the PVC is used by a pod: adds the PVC into a set of PVCs that are in use and are waiting for deletion.
-
-PVC deletion controller reacts to pod termination in the below way:
-- In case the pod was using PVC(s) that are waiting for deletion and these PVC(s) are not used by any other pods: the PVC deletion controller live updates pods in the PVC's namespace and in case the PVC is not used by any pods it removes the finalizer information from the PVC metadata.
-
-Note: there may be PVC(s) without `deletionTimestamp` set and without finalizer information set after Kubernetes upgrade.
-
-That's why the PVC deletion controller will upon PVC create or update:
-- Add finalizer information to all PVCs that don't have the `deletionTimestamp` set and don't have the finalizer information set.
-
-Note: the PVC deletion controller may be down for some time.
-
-PVC deletion controller after it starts gets current status of pods and PVCs from API server. Afterwards, it does the below:
-- Removes the finalizer information from the PVC metadata for all PVC that have `deletionTimestamp` set and are not used by a pod.
-- Adds all PVCs that have the `deletionTimestamp` set and are used by a pod into a set of PVCs that are in use and are waiting for deletion.
+PVC deletion controller watches for both PVC and pod events that are processed as described below:
+1. PVC add/update/delete events:
+  - If `deletionTimestamp` is `nil` and finalizer is missing, the PVC is added to PVC queue.
+  - If `deletionTimestamp` is `non-nil` and finalizer is present and first, the PVC is added to PVC queue.
```

Comment:
Why the finalizer position matters?

Reply:
you're right, I was thinking finalizers were handled in order, but that is not the case. @pospispa you can strike " and first"